oceannavigator/frontend/src/components/Map/Map.jsx

```diff
@@ -899,26 +907,98 @@ const Map = forwardRef((props, ref) => {
+  const addAnnotationLabel = (text, coord) => {
+    const el = document.createElement("div");
+    el.className = "annotation-box";
+    const closeBtn = document.createElement("span");
+    closeBtn.className = "annotation-close";
+    closeBtn.innerHTML = "×";
+    el.appendChild(closeBtn);
+    const textNode = document.createTextNode(text);
+    el.insertBefore(textNode, closeBtn);
+
+    const overlay = new Overlay({
+      element: el,
+      position: coord,
+      positioning: "bottom-left",
+      offset: [0, -30],
+    });
+    map0.addOverlay(overlay);
+    setAnnotationOverlays((prev) => [...prev, overlay]);
+    let dragging = false;
+    let startPixel;
+
+    el.addEventListener("mousedown", (evt) => {
+      dragging = true;
+      startPixel = [evt.clientX, evt.clientY];
+      evt.stopPropagation();
+    });
+
+    window.addEventListener("mousemove", (evt) => {
+      if (!dragging) return;
+      const deltaX = evt.clientX - startPixel[0];
+      const deltaY = evt.clientY - startPixel[1];
+      startPixel = [evt.clientX, evt.clientY];
+
+      const currPixel = map0.getPixelFromCoordinate(overlay.getPosition());
+      const newPixel = [currPixel[0] + deltaX, currPixel[1] + deltaY];
+      const newCoord = map0.getCoordinateFromPixel(newPixel);
+      overlay.setPosition(newCoord);
+    });
+
+    window.addEventListener("mouseup", () => {
+      dragging = false;
+    });
+
+    closeBtn.addEventListener("click", (evt) => {
+      map0.removeOverlay(overlay);
+      setAnnotationOverlays((all) => all.filter((o) => o !== overlay));
+      evt.stopPropagation();
     });
-    annotationVectorSource.addFeature(feature);
   };
 
   const undoAnnotationLabel = () => {
-    let features = annotationVectorSource.getFeatures();
-    if (features.length > 0) {
-      annotationVectorSource.removeFeatures([features[features.length - 1]]);
-    }
+    setAnnotationOverlays((all) => {
+      const list = [...all];
+      const last = list.pop();
+      if (last) map0.removeOverlay(last);
+      return list;
+    });
   };
 
   const clearAnnotationLabels = () => {
-    annotationVectorSource.clear();
+    annotationOverlays.forEach((o) => map0.removeOverlay(o));
+    setAnnotationOverlays([]);
+  };
+
+  const enableAnnotationMode = () => {
+    setAnnotationMode(true);
+    map0.getViewport().style.cursor = "crosshair";
+  };
+  const disableAnnotationMode = () => {
+    setAnnotationMode(false);
+    map0.getViewport().style.cursor = "";
   };
 
+  useEffect(() => {
+    if (!map0) return;
+    const onMapClick = (e) => {
+      if (!annotationMode) return;
+      const coord = e.coordinate;
+      disableAnnotationMode();
+      props.updateAnnotationCoord(coord);
+      props.updateUI({
+        annotationMode: false,
+        modalType: "annotation",
+        showModal: true,
+      });
+    };
+    map0.on("click", onMapClick);
+    return () => {
+      map0.un("click", onMapClick);
+    };
+  }, [map0, annotationMode]);
+
```

Comment:
Instead of using an effect hook here to add the click listener you can just add the click listener to the map object when it gets created like we do here: https://github.com/DFO-Ocean-Navigator/Ocean-Data-Map-Project/blob/3833ccfbeed55a1710a4341156fa52d5d23aba4e/oceannavigator/frontend/src/components/Map/Map.jsx#L219